Support multiple preview paths for mailers

fatkodima opened the pull request:
It should be great to see mailer previews from engines. This PR allows to do this.

Originated from a bit outdated https://github.com/rails/rails/pull/16836.
Notable changes compared to that PR:
- `config.action_mailer.preview_path` deprecating
- Engine initializer to automatically add mailer preview paths to the lookup path for mailer previews for the application

Thanks in advance.

fatkodima committed:
Support multiple preview paths for mailers

georgeclaghorn commented on actionmailer/lib/action_mailer/preview.rb:
```diff
@@ -25,7 +25,31 @@ module Previews #:nodoc:
+        Please use preview_paths instead.
+      MSG
+      self.class.preview_paths.first
+    end
```
This instance method is currently private, so I think it can be removed without deprecation.

fatkodima replied:
Looks like instance reader is public https://github.com/rails/rails/blob/bce675eac49a8673d63ea41991f020aa3aa8c506/actionmailer/lib/action_mailer/preview.rb#L14

georgeclaghorn replied:
You‘re right, sorry. 😅

georgeclaghorn commented on railties/test/application/mailer_previews_test.rb:
```diff
@@ -253,8 +253,70 @@ def foo
+      assert_match '<li><a href="/rails/mailers/notifier/foo">foo</a></li>', last_response.body
+      assert_match '<h3><a href="/rails/mailers/confirm">Confirm</a></h3>', last_response.body
+      assert_match '<li><a href="/rails/mailers/confirm/foo">foo</a></li>', last_response.body
+    end
```
This test can be combined with the existing one for loading previews from custom preview paths above.

georgeclaghorn reviewed (changes requested):
:heart:

fatkodima commented:
@georgeclaghorn, thanks for review. Fixed your notes, squashed and rebased.
Anything else?

MadLittleMods commented:
Friendly ping to revive 👋 

We have a use case for this to support some GitLab Enterprise Edition specific email previews which are namespaced under the `ee/` directory.

```rb
config.action_mailer.preview_paths = [
  'spec/mailers/previews',
  'ee/spec/mailers/previews'
]
```

---

Here is the current config, https://gitlab.com/gitlab-org/gitlab-ce/blob/b17457a7f7fb390c829ff5feed7e4635bf2b4521/config/environments/development.rb#L42

fatkodima commented:
@georgeclaghorn Wdyt about moving this forward?

simi commented:
I was just looking for the implementation details. Since it pass `preview_path` to `Dir[]`, isn't is possible to pass multiple values in matching literal `{a,b}? 

```ruby
ActionMailer::Base.preview_path = "{/etc,/tmp}"
```

Anyway it is just non-intuitive workaround. Introducing `action_mailer.preview_paths` seems more intuitive.

simi commented on actionmailer/lib/action_mailer/preview.rb:
```diff
@@ -25,7 +25,31 @@ module Previews #:nodoc:
 
+    def preview_path
+      ActiveSupport::Deprecation.warn(<<-MSG.squish)
+        Using preview_path option is deprecated and will be removed in Rails 6.0.
```
:hourglass: Rails 6.1 now

fatkodima commented:
@eugeneius please, review 🙏

eugeneius commented on actionmailer/lib/action_mailer/railtie.rb:
```diff
@@ -25,7 +26,9 @@ class Railtie < Rails::Railtie # :nodoc:
-        options.preview_path ||= defined?(Rails.root) ? "#{Rails.root}/test/mailers/previews" : nil
+        options.preview_paths |= ["#{Rails.root}/test/mailers/previews"]
+      else
+        options.preview_paths.clear
```
What's this for? It doesn't look like it's covered by tests.

fatkodima replied:
👍 Yes, this is actually not needed.

eugeneius commented on actionmailer/lib/action_mailer/railtie.rb:
```diff
@@ -57,8 +60,10 @@ class Railtie < Rails::Railtie # :nodoc:
-      if options.show_previews && options.preview_path
-        ActiveSupport::Dependencies.autoload_paths << options.preview_path
+      if options.show_previews
+        options.preview_paths.each do |preview_path|
+          ActiveSupport::Dependencies.autoload_paths << preview_path
+        end
```
```suggestion
        ActiveSupport::Dependencies.autoload_paths.concat(options.preview_paths)
```

eugeneius commented on guides/source/action_mailer_basics.md:
````diff
@@ -518,12 +518,12 @@ visually see the new style instantly. A list of previews are also available
 
 ```ruby
-config.action_mailer.preview_path = "#{Rails.root}/lib/mailer_previews"
+config.action_mailer.preview_paths << "#{Rails.root}/lib/mailer_previews
````
```suggestion
config.action_mailer.preview_paths << "#{Rails.root}/lib/mailer_previews"
```

eugeneius commented on railties/test/application/mailer_previews_test.rb:
```diff
@@ -254,9 +258,31 @@ def foo
-    test "mailer previews are reloaded from a custom preview_path" do
+    test "mailer preview_path option is deprecated" do
+      prev = ActionMailer::Base.preview_paths
+      ActiveSupport::Deprecation.silence do
+        ActionMailer::Base.preview_path = "#{app_path}/lib/mailer/previews"
+      end
```
```suggestion
      ActionMailer::Base.preview_paths = ["#{app_path}/lib/mailer/previews"]
```

eugeneius commented:
> Engine initializer to automatically add mailer preview paths to the lookup path for mailer previews for the application

Did this part not work out? It sounds useful, since that's the main motivating use case for supporting multiple preview paths.

kaspth commented:
Yeah that, and engines auto include controllers/routes etc in the main app. So this would fit right with that.

fatkodima commented:
@eugeneius thanks! Updated.

Re:
> Engine initializer to automatically add mailer preview paths to the lookup path for mailer previews for the application

I initially implemented that. Then something went wrong and I decided to remove this and allow for engines to manually add preview paths. And now everything went right ( 😄 )and I added back engine initializer (and tests for it).

eugeneius commented on railties/lib/rails/engine.rb:
```diff
@@ -618,6 +618,13 @@ def load_seed
     end
 
+    initializer :add_mailer_preview_paths do
+      previews = paths["test/mailers/previews"].existent
```
The contents of `config.action_mailer.preview_paths` are added to `ActiveSupport::Dependencies.autoload_paths`, I think we should do the same with these paths too.

It's not necessary for the mailer previews to work, but it will make a naming conflict between a preview and another autoloaded constant more apparent, and I suspect the way previews are currently eager loaded ([`require_dependency` with absolute paths](https://github.com/rails/rails/blob/db543ba728a9bd38a08f62ebb93ec9eca802a987/actionmailer/lib/action_mailer/preview.rb#L123)) won't work once the classic autoloader is deprecated.

fatkodima replied:
Updated `paths.add "test/mailers/previews", eager_load: true` to serve this purpose. This should work, I believe?

eugeneius replied:
Yep, that's even better 👍 

I wonder if the railtie should add its preview paths to `paths["test/mailers/previews"]` instead of directly manipulating `ActiveSupport::Dependencies.autoload_paths`?

```patch
diff --git a/actionmailer/lib/action_mailer/railtie.rb b/actionmailer/lib/action_mailer/railtie.rb
index 6bae8c825e..2bd5342fca 100644
--- a/actionmailer/lib/action_mailer/railtie.rb
+++ b/actionmailer/lib/action_mailer/railtie.rb
@@ -55,12 +55,10 @@ class Railtie < Rails::Railtie # :nodoc:
       end
     end

-    initializer "action_mailer.set_autoload_paths" do |app|
+    initializer "action_mailer.set_autoload_paths", before: :set_autoload_paths do |app|
       options = app.config.action_mailer

-      if options.show_previews
-        ActiveSupport::Dependencies.autoload_paths.concat(options.preview_paths)
-      end
+      app.config.paths["test/mailers/previews"].concat(options.preview_paths)
     end

     initializer "action_mailer.compile_config_methods" do
```

eugeneius commented on railties/lib/rails/engine.rb:
```diff
@@ -618,6 +618,13 @@ def load_seed
 
+    initializer :add_mailer_preview_paths do
+      previews = paths["test/mailers/previews"].existent
+      unless previews.empty?
```
The Action Mailer railtie doesn't add these paths if `show_previews` is false. Should we do the same here?

Is the `empty?` check purely an optimisation? If so, maybe it makes sense to add it in the railtie, too.

fatkodima replied:
> The Action Mailer railtie doesn't add these paths if show_previews is false. Should we do the same here?

I think this is not necessary. We are not forced to keep `preview_paths` empty when `show_previews == false`, because previews just won't be showed in `/rails/mailers` by that setting. But avoiding that check (which should be in `on_load` block due to that we need `ActionMailer` loaded at that time) we simplified code a bit.

> Is the empty? check purely an optimisation? If so, maybe it makes sense to add it in the railtie, too.

Yes, it is some sort of unnecessary optimization, similarly as in `:add_routing_paths` and `:add_view_paths` initializers above. I think this is also not needed for railtie (to simplify code a bit, like in previous comment).

eugeneius replied:
Good point about the `on_load` hook - it won't even work in development, since it doesn't fire until the first time a mailer is used, at which point the autoload paths are already frozen.

I'd like to avoid the situation where `preview_paths` contains paths from engines but not from the application, so I think we should take the `show_previews` check out of the railtie instead.

eugeneius commented on railties/test/application/paths_test.rb:
```diff
@@ -53,6 +53,8 @@ def assert_not_in_load_path(*path)
       assert_path @paths["config/routes.rb"],    "config/routes.rb"
       assert_path @paths["config/routes"],       "config/routes"
+      assert_path @paths["test/mailers/previews"], "test/mailers/previews"
+
```
Can you ✂️ the extra newline here? 😁

fatkodima commented:
Updated.

fatkodima commented:
@eugeneius any chance to get this merged?

kennyadsl commented:
This would be super useful for Rails Engines that add mailer previews in the host application as we do in Solidus. We worked around the issue with [this PR](https://github.com/solidusio/solidus/pull/3901) but would be great to have support for that in Rails directly. Any chance to move this forward?

p8 commented:
@fatkodima could you fix the merge conflicts and update the deprecations from `6.2` to `7.1`? 🙇

fatkodima commented:
@p8 Rebased and changed `6.2` to `7.1`.

p8 commented:
Thanks @fatkodima !

fxn commented:
I believe this needs a revision.

Since `config.paths` is cached as soon as accessed, `config.autoload_paths` and friends are going to have whatever `paths["test/mailers/previews"]` had at the time the object was memoized.

Therefore, concatenating to `paths["test/mailers/previews"]` does not necessarily add them to the autoload paths, it depends on whether the user or engines accessed `config.paths` in any way (like setting `config.autoload_paths`). If they did, mailer previews won't be found.

Also, if I am not mistaken, only applications can configure mailer previews via `config.action_mailer.preview_paths`, so the flexibility at engine level is unnecessary.

I would consider removing `paths["test/mailers/previews"]` and having the initializer push things to the autoload paths by concatenating into `config.autoload_paths` by hand, which is good interface between railties and engines or applications. Please, see https://github.com/rails/rails/pull/49032, which needed a workaround for this.

fxn commented:
Let me illustrate the regression.

In Rails 7.0, if you have a directory `lib/mailer_previews` and this in `config/application.rb`:

```ruby
config.autoload_paths
config.action_mailer.preview_path = "#{Rails.root}/lib/mailer_previews"
```

then

```
% bin/rails r 'pp ActiveSupport::Dependencies.autoload_paths.grep(/lib/)'
["/Users/fxn/tmp/del_mailer_previews/lib/mailer_previews"]
```

In `main`, `lib/mailer_previews` is not in the autoload paths.

Besides being a regression, conceptually this does not look good. It wouldn't be right even if the feature was new.

fatkodima commented:
Please, feel free to provide a PR, if you will, since you better know how to implement it properly. 🙏

fxn commented:
@fatkodima let me understand your patch better. The initializer

```ruby
initializer "action_mailer.set_autoload_paths", before: :set_autoload_paths do |app|
  options = app.config.action_mailer
  app.config.paths["test/mailers/previews"].concat(options.preview_paths)
end
```

adds preview paths from the application/global configuration point `app.config.action_mailer.preview_paths`. In what sense does this patch allow to configure these per engine too?

fatkodima commented:
These changes https://github.com/rails/rails/pull/31595/files#diff-c72ea90e9c6b7c60285006ea4c34548ae9257dffcda0c1b2f41fffd75a14363b are expected to do that.

fxn commented:
So engines do not have configuration API for it, right? (I don't consider `config.paths["..."]` to be public API).

fatkodima commented:
Right.

jon-sully commented:
For those coming here because their Previews no longer got autoloaded after upgrading to Rails 7.1 and that use RSpec (so your previews are under `/spec...` not `/test...`), RSpec isn't updated to inject the `preview_paths` correctly quite yet:

https://github.com/rspec/rspec-rails/issues/2703

In the meantime, 

```ruby
  config.action_mailer.preview_paths << "#{Rails.root}/spec/mailers/previews"
```
